src/Bicep.Core/Syntax/SyntaxTreeGroupingBuilder.cs

```diff
@@ -146,6 +146,55 @@ private SyntaxTree AddSyntaxTree(Uri fileUri, string fileContents)
             return syntaxTree;
         }
 
+        private static readonly ImmutableHashSet<char> forbiddenPathChars = "<>:\"\\|?*".ToImmutableHashSet();
+        private static readonly ImmutableHashSet<char> forbiddenPathTerminatorChars = " .".ToImmutableHashSet();
+
+        private bool ValidateModulePath(string pathName, [NotNullWhen(false)] out DiagnosticBuilder.ErrorBuilderDelegate? failureBuilder)
+        {
+            if (pathName.Length == 0)
+            {
+                failureBuilder = x => x.ModulePathIsEmpty();
+                return false;
+            }
+
+            if (pathName.First() == '/')
+            {
+                failureBuilder = x => x.ModulePathBeginsWithForwardSlash();
+                return false;
+            }
+
+            foreach (var pathChar in pathName)
+            {
+                if (pathChar == '\\')
+                {
+                    // enforce '/' rather than '\' for module paths for cross-platform compatibility
+                    failureBuilder = x => x.ModulePathContainsBackSlash();
+                    return false;
+                }
+
+                if (forbiddenPathChars.Contains(pathChar))
+                {
+                    failureBuilder = x => x.ModulePathContainsForbiddenCharacters(forbiddenPathChars);
+                    return false;
+                }
+
+                if (pathChar >= 0 && pathChar <= 31)
```

Comment:
>pathChar >= 0 && pathChar <= 31 [](start = 20, length = 31)

Normally, I'd also ask if we should just call char.IsControl() ([https://docs.microsoft.com/en-us/dotnet/api/system.char.iscontrol?view=netcore-3.1](https://docs.microsoft.com/en-us/dotnet/api/system.char.iscontrol?view=netcore-3.1)), but I have two conflicting thoughts about it:
1. This is a parser, so we need to control what characters are allowed where, so that could mean unexpected changes if dotnet updates its Unicode implementation.
2. When we add Unicode support to the language, we will likely have to use dotnet's implementation, which means reusing stuff like this or at least duplicating the logic.

Reply:
Could you add a comment like "// TODO: Revisit when adding Unicode", so it's easier to find later?

---
In reply to: [516204313](https://github.com/Azure/bicep/pull/766#discussion_r516204313) [](ancestors = 516204313)